lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,443 @@
+LIBID = "196ff8f539ba4f2998209fbb50e2dbbf"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+"""Schemas definition for the provider and requirer sides of the `fiveg_core_gnb` interface.
+It exposes two interfaces.schema_base.DataBagSchema subclasses called:
+- ProviderSchema
+- RequirerSchema
+
+Examples:
+    ProviderSchema:
+        unit: <empty>
+        app: {
+            "tac": 1,
+            "plmns": [
+                {
+                    "mcc": "001",
+                    "mnc": "01",
+                    "sst": 1,
+                    "sd": 1,
+                }
+            ],
+        }
+    RequirerSchema:
+        unit: <empty>
+        app: {
+            "cu_name": "gnb001",
+        }
+"""
+
+
+@dataclass
+class PLMNConfig(BaseModel):
+    """Dataclass representing the configuration for a PLMN."""
+
+    def __init__(self, mcc: str, mnc: str, sst: int, sd: Optional[int] = None) -> None:
+        super().__init__(mcc=mcc, mnc=mnc, sst=sst, sd=sd)
+
+    mcc: str = Field(
+        description="Mobile Country Code",
+        examples=["001", "208", "302"],
+        pattern=r"[0-9][0-9][0-9]",
+    )
+    mnc: str = Field(
+        description="Mobile Network Code",
+        examples=["01", "001", "999"],
+        pattern=r"[0-9][0-9][0-9]?",
+    )
+    sst: int = Field(
+        description="Slice/Service Type",
+        strict=True,
+        examples=[1, 2, 3, 4],
+        ge=0,
+        le=255,
+    )
+    sd: Optional[int] = Field(
+        description="Slice Differentiator",
+        strict=True,
+        default=None,
+        examples=[1],
+        ge=0,
+        le=16777215,
+    )
+
+    def asdict(self):
+        """Convert the dataclass into a dictionary."""
+        return {"mcc": self.mcc, "mnc": self.mnc, "sst": self.sst, "sd": self.sd}
+
+
+class FivegCoreGnbProviderAppData(BaseModel):
+    """Provider application data for fiveg_core_gnb."""
+    tac: int = Field(
+        description="Tracking Area Code",
+        strict=True,
+        examples=[1],
+        ge=1,
+        le=16777215,
+    )
+    plmns: list[PLMNConfig]
+
+
+class ProviderSchema(DataBagSchema):
+    """Provider schema for fiveg_core_gnb."""
+
+    app_data: FivegCoreGnbProviderAppData
+
+
+def data_matches_provider_schema(data: dict) -> bool:
+    """Return whether data matches provider schema.
+
+    Args:
+        data (dict): Data to be validated.
+
+    Returns:
+        bool: True if data matches provider schema, False otherwise.
+    """
+    try:
+        ProviderSchema(app_data=FivegCoreGnbProviderAppData(**data))
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
+
+class GnbAvailableEvent(EventBase):
```

Comment:
While discussing with @patriciareinoso Today, I realised that we don't need custom events anymore at all, now that we use the central hook approach.

Instead of emitting this event, and observing it in the charm code, we can implement only getters and setters here in the lib and use those in the _configure method in charm.py.

Getting rid of custom events would simplify our libs quite a bit.

I don't know that we need to go and change all of our libs to get rid of them, but at least for new libs like this one, I would not go and add this unnecessary complexity.